File: python_client/board_and_history.py
import ffi
from ffi import * 
import random

class Bot:
    def __init__(self, log):
        pass

    def take_action(self, game_info, log):
        # There's lots of useful information in the game_info object!
        num_players = game_info.num_players
        log.send(f"Explanation of a game with {num_players} players")

        # Game information about the current state and the 
        # history of the game until the current point can be accessed
        # like so:

        # You can iterate over the previous states of the game
        history = game_info.history
        all_turns = history.turns
        last_5_turns = all_turns[-5:]

        # You can see what actions were taken during those turns
        for (player_index, actions) in last_5_turns:
            for action in actions:
                if action.action_type == PyActionType.AttractNoble:
                    log.send( f"Careful! Player {player_index} attracted a noble within the last 5 turns!!")
                    log.send( f"They might be close to winning!")

        # You can also see the current state of the game by accessing the board

        board = game_info.board
        # The board has the following attributes:
        board.deck_counts[0] # The number of tier 0 cards face down in the deck
        board.deck_counts[1] # The number of tier 1 cards face down in the deck
        board.deck_counts[2] # The number of tier 2 cards face down in the deck

        board.nobles # The list of nobles that are still available for grabs

        # Or you can check out the cards that are still available
        board.face_up_cards( tier = 0 )
        board.face_up_cards( tier = 1 )
        board.face_up_cards( tier = 2 )

        # Or how many of what type of gem is left in the bank
        board.gems.onyx
        board.gems.sapphire
        board.gems.diamond
        board.gems.emerald
        board.gems.ruby
        board.gems.gold


        if len(board.nobles) > 0:
            log.send(f"There are {len(board.nobles)} nobles left")
            noble = board.nobles[0]

            # You can access the noble's cost
            noble.cost.onyx
            noble.cost.sapphire
            noble.cost.diamond
            noble.cost.emerald
            noble.cost.ruby

            # Or access the noble's id
            noble.id

            # Or create a noble by id
            if PyNoble( noble.id ) == noble:
                log.send("All is right in the noble world...")


        # You can also see the current state of the game by accessing the players
        players = game_info.players
        my_player_index = game_info.me.index
        player = players[my_player_index]

        # You can access the player's various properties
        player.index  # The index of the player (0 up to num_players - 1)

        player.total_points # The number of points the player has scored (including points from nobles)

        player.gems   # The breakdown of gems they have
        player.gems.onyx
        player.gems.sapphire
        player.gems.diamond
        player.gems.emerald
        player.gems.ruby
        player.gems.gold

        player.num_reserved_cards # The number of cards they have reserved 

        player.developments  # The breakdown of the gem type on cards they have played
        player.developments.onyx 
        player.developments.sapphire
        player.developments.diamond
        player.developments.emerald
        player.developments.ruby


        legal_actions = game_info.legal_actions
        action = random.choice(legal_actions)
        return action

if __name__ == "__main__":
    ffi.run_python_bot(Bot)

File: rust_client/src/examples/board_and_history.rs
use crate::stubs::*;

#[derive(Debug, Default)]
pub struct Bot {
}

impl Runnable<GameInfo, Action> for Bot {
    fn initialize(&mut self, _log: &mut Log) {
        // Initialize your bot here!
    }

    fn take_action(&mut self, info: GameInfo, log: &mut Log) -> Action {

        // There's lots of useful information in the GameInfo object!
        let _num_players = info.num_players;
        log.send("[board_and_history] hello from the bot");

        // Game information about the current state and the 
        // history of the game until the current point can be accessed
        // like so:

        // You can iterate over the previous states of the game
        let history = &info.history;
        let all_turns = &history.turns;
        let last_5_turns = &all_turns[all_turns.len() - 5..];

        for (player_index, actions) in last_5_turns.iter() {
            for action in actions {
                if matches!(action, Action::AttractNoble(_)) {
                    let warning = format!("[board_and_history] player {} attracted a noble", player_index);
                    let advice = "[board_and_history] maybe they're close to winning?";
                    log.send(&warning);
                    log.send(advice);
                }
            }
        }

         
        // You can also see the current state of the game by accessing the board
        let board = &info.board;
        // The board has the following attributes:
        board.deck_counts[0]; // The number of tier 0 cards left in the deck
        board.deck_counts[1]; // The number of tier 1 cards left in the deck
        board.deck_counts[2]; // The number of tier 2 cards left in the deck
        
        let _nobles = board.nobles.clone(); // The list of noble ids that are still available to grab
        let _noble_0 = Noble::from_id(0); 
        let _noble_1 = Noble::from_id(1); 
        let _noble_5 = Noble::from_id(5); 

        // You can see how many of what type of gem is left in the bank
        board.gems.onyx;
        board.gems.gold;

        // You can also see the current state of the game by accessing the players
        let players = &info.players;

        let _ = players[0].reserved_cards.clone(); // The cards that player 0 has reserved
        // Which may be None if you are not able to see the player's reserved cards

        // You can also access various properties of each player
        let my_player = &info.me();
        my_player.gems; // The number of gems the player has
        my_player.total_points; // The number of points the player has (including from nobles)
        my_player.developments; // The cards the player has played in front of them
        my_player.num_reserved_cards; // The number of cards the player currently has reserved
        

        let legal_actions = info.legal_actions;
        legal_actions[0].clone()
    }
}
